### PR TITLE
One Q/A is based on a single tabular dataset and the answer is useful

### LINKED ISSUE #5: At least one Q&A is based on a single (tabular) dataset, and the answer is useful

### PR DESCRIPTION
**Pull request by d-geula**
lanchain_exmaple_2
lanchain_exmaple_1

### CODE REVIEW

File: app.py [hunk 1]
```diff
@@ -13,29 +12,23 @@
 
 
 # Main function
-def nl_to_sql(prompt):
-    # Translate the natural language query to SQL using the OpenAI API
-    sql_query = utils.get_openai_response(prompt, "sql_agent")
-
-    # Process the SQL query
-    query_result = utils.process_query(sql_query)
+def nl_to_sql(prompt: str):
```
Comment (itaifrenkel) [inline app.py]: change this method (or refactor to a _private method) such that it accepts `llm: langchain.llms.base.LLM`
And in the "production" code make sure to provide it with an OpenAI object, while in the unit testing code use https://github.com/hwchase17/langchain/blob/master/langchain/llms/fake.py

This is what we call a stub (or fake ://)

File: app.py [hunk 2]
```diff
@@ -1,27 +1,29 @@
...
+    db_path = "db/formula1.sqlite"
+    db = SQLDatabase.from_uri(f"sqlite:///{db_path}", sample_rows_in_table_info=3)
 
-# Main function
-def nl_to_sql(prompt: str):
-    db_path = "files/nba.sqlite"
-    db = SQLDatabase.from_uri(f"sqlite:///{db_path}", sample_rows_in_table_info=2)
-    llm = OpenAI(temperature=0, verbose=False)  # type: ignore
+    if not _test:
+        llm = ChatOpenAI(model="gpt-3.5-turbo", temperature=0, verbose=True)  # type: ignore
```
Comment (itaifrenkel) [inline app.py]: This code could move to nl_to_sql_chatgpt function that calls nl_to_sql

Then move the stubs to the test function in the test file.
the test covers nl_to_sql function but not chatgpt function. That’s the nature of testing with stubs.

the reason is_test is not a best practice is that it mixes test code (stubs) inside production code, and adds complexity (a single if statement in this case)

WDYT?

**Comment by d-geula** [comment]
1. While the test for the db chain is actually doing something now, I'm not really sure exactly what I should even be testing there so let me know if I need to modify it. It passes btw.
2. I hope this isn't too much of an issue but this branch now also includes the code for this issue: https://github.com/d-geula/NL2SQL-Translator/issues/10
Because I did it all together on a separate test branch, and because I haven't even opened a new branch for that issue, I thought it would just be easier to put it all here.

**Review by itaifrenkel — APPROVED** [review]

**Comment by itaifrenkel** [comment]
I merged the PR. Please mark relevant project items as done .

For now the logic other than the stub is very trivial  so the test is not useful. Leave it for when you start improving the logic

Regards,
Itai
________________________________

**Pull request merged**